account_invoice_validation_queued/wizards/account_invoice_confirm.py

```diff
@@ -0,0 +1,35 @@
+# Copyright 2019 Tecnativa - Pedro M. Baeza
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, exceptions, models
+
+
+class AccountInvoiceConfirm(models.TransientModel):
+    _inherit = "account.invoice.confirm"
+
+    def enqueue_invoice_confirm(self):
+        queue_obj = self.env['queue.job']
+        active_ids = self.env.context.get('active_ids', [])
+        invoices = self.env['account.invoice'].browse(active_ids)
+        if invoices.filtered(lambda x: x.state not in 'draft'):
+            # Let standard method to raise the exception about draft state
+            return self.invoice_confirm()
+        # Check if all invoices are for the same date
+        inv_date = invoices[:1].date_invoice
+        for invoice in invoices:
+            if invoice.date_invoice != inv_date:
+                raise exceptions.UserError(_(
+                    "You can't enqueue invoices with different dates."
+                ))
+        for invoice in invoices:
+            if invoice.validation_job_ids.filtered(
```

Comment:
@pedrobaeza Queue job allows you to define a job_identity key to avoid the creation of duplicates. No need of specific code here... https://github.com/OCA/queue/commit/2e86357819315e82bba12a39149ed82614f67d8b

Reply:
But this is not unique... You can validate the invoice several times (you cancel it and validate it again).

Reply:
![image](https://user-images.githubusercontent.com/544090/64533026-f97b7300-d312-11e9-8fe1-aeb7f5d4c366.png)

Reply:
OK, good to know. I will change this in a new PR when we clarify all the comments.

Reply:
Also, jobs must be idempotent: the first thing they should to is verify the state and return if it doesn't match its expectation (e.g. return if the state is already 'open').

By removing this custom code, you will be able to remove the override of `action_cancel()` that you added on `queue.job`. I agree with @lmignon that it shouldn't be there. You shouldn't change the behavior of the `queue.job` model here.

Reply:
The logic for having that method is that `unlink` is already public, so this is not a backdoor, but it's a safe method for not needing to worry about which jobs can be removed in this context, but if it's so problematic, I'll remove it and call `unlink` directly. That's not a big problem at all.

@guewen what about changing the user that will execute the job. Is it possible from the `with_delay` call?

Reply:
About the idempotent you mentioned, @guewen, I have double check again, and the method that is call from the job is already handling the cases where the state is different from `open`, and that's why I prefer to leave them that way, so that the job fails in that case with the exception raised from upstream.

Reply:
>   what about changing the user that will execute the job. Is it possible from the with_delay call?

The job is executed as the user delaying the job. However, the user does not need access rights on `queue.job` because `with_delay()` creates the record using `sudo()`.



>  I have double checked again, and the method that is call from the job is already handling the cases where the state is different from `open`, and that's why I prefer to leave them that way, so that the job fails in that case with the exception raised from upstream.

If the state is already good, the job has nothing to do: it should be `done`.

Reply:
> The job is executed as the user delaying the job. However, the user does not need access rights on queue.job because with_delay() creates the record using sudo().

Great then, changing the code for removing `sudo`.

> If the state is already good, the job has nothing to do: it should be done.

Well, Odoo thinks the contrary on this, as it raises the exception if you are trying to validate again one validated invoice. What do you think about this, @lmignon ?

Reply:
> Well, Odoo thinks the contrary on this, as it raises the exception if you are trying to validate again one validated invoice

My point is not about odoo's code but about jobs. Jobs are not user-oriented, they are background, low-level representation of "a method running a background" (think celery). Users should not have to deal with them and have to care about them (beside a few advanced users tracking issues). A good functioning system using jobs should have *zero* jobs failing (in practice it's inevitable to have some but that should be the goal). The thing is you modified the UI to show the job records to users, that's why you have different expectations, but IMO that's a bad idea. The job should only validate the invoice, if it's already validated then it should return and be done; if there is an error during validation, the job should log a message in the invoice's thread and be done. But "normal" users should not have to deal with failed jobs.

Reply:
We have different vision about this and what the user should know and interact with, but in this case, as the scope is very limited, I'm adding the condition about the state in https://github.com/OCA/account-invoicing/compare/fc5af5be55fcb243fa7ef1500be236415c264959..85444ce0e5d09adcf335059db1f9a78013f50eb4 for #593 for not having the job failing.